plugins/amazonq/chat/jetbrains-community/src/software/aws/toolkits/jetbrains/services/cwc/inline/InlineChatController.kt

```diff
@@ -0,0 +1,722 @@
+import software.aws.toolkits.jetbrains.services.cwc.controller.chat.userIntent.UserIntentRecognizer
+import software.aws.toolkits.jetbrains.services.cwc.editor.context.ActiveFileContextExtractor
+import software.aws.toolkits.jetbrains.services.cwc.editor.context.ExtractionTriggerType
+import software.aws.toolkits.jetbrains.services.cwc.inline.listeners.InlineChatFileListener
+import software.aws.toolkits.jetbrains.services.cwc.messages.ChatMessage
+import software.aws.toolkits.jetbrains.services.cwc.messages.ChatMessageType
+import software.aws.toolkits.jetbrains.services.cwc.storage.ChatSessionStorage
+import software.aws.toolkits.jetbrains.settings.CodeWhispererSettings
+import software.aws.toolkits.telemetry.FeatureId
+import java.util.UUID
+import java.util.concurrent.atomic.AtomicBoolean
+import kotlin.math.max
+
+@Service(Service.Level.PROJECT)
+class InlineChatController(
+    private val project: Project,
+    private val scope: CoroutineScope,
+) : Disposable {
+    private var currentPopup: JBPopup? = null
+    private var rangeHighlighter: RangeHighlighter? = null
+    private var rejectAction: (() -> Unit)? = null
+    private var acceptAction: (() -> Unit)? = null
+    private var insertionLine = AtomicInteger(-1)
+    private val sessionStorage = ChatSessionStorage()
+    private val telemetryHelper = TelemetryHelper(project, sessionStorage)
+    private val shouldShowActions = AtomicBoolean(false)
+    private val isInProgress = AtomicBoolean(false)
+    private var metrics: InlineChatMetrics? = null
+    private var canPopupAbort = AtomicBoolean(true)
+    private var currentSelectionRange: RangeMarker? = null
+    private val listener = InlineChatFileListener(project, this)
+    private var isAbandoned = AtomicBoolean(false)
+
+    init {
+        Disposer.register(this, listener)
+        project.messageBus.connect(this).subscribe(FileEditorManagerListener.FILE_EDITOR_MANAGER, listener)
+    }
+
+    data class InlineChatMetrics(
+        val requestId: String,
+        val inputLength: Int? = null,
+        val numSelectedLines: Int? = null,
+        val codeIntent: Boolean? = null,
+        var userDecision: InlineChatUserDecision? = null,
+        val responseStartLatency: Double? = null,
+        val responseEndLatency: Double? = null,
+        var numSuggestionAddChars: Int? = null,
+        var numSuggestionAddLines: Int? = null,
+        var numSuggestionDelChars: Int? = null,
+        var numSuggestionDelLines: Int? = null,
+        var programmingLanguage: String? = null,
+    )
+
+    private val popupSubmitHandler: suspend (String, String, Int, Editor) -> String = {
+            prompt: String, selectedCode: String, selectedLineStart: Int, editor: Editor ->
+        runBlocking {
+            isInProgress.set(true)
+            val message = handleChat(prompt, selectedCode, editor, selectedLineStart)
+            message
+        }
+    }
```

Comment:
```suggestion
    private val popupSubmitHandler: suspend (String, String, Int, Editor) -> String = {
            prompt, selectedCode, selectedLineStart, editor ->
            isInProgress.set(true)
            val message = handleChat(prompt, selectedCode, editor, selectedLineStart)
            message
    }
```

or

```suggestion
        private suspend fun popupSubmitHandler(prompt: String, selectedCode: String, selectedLineStart: Int, editor: Editor): String {
            isInProgress.set(true)
            val message = handleChat(prompt, selectedCode, editor, selectedLineStart)
            return message
        }
```

and then you should be able to do
```
InlineChatPopupFactory(
            submitHandler = ::popupSubmitHandler,
        )
```